nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,1095 @@
+        for p in self:
+            if isinstance(p, Cifti2Volume):
+                return p
+        return None
+
+    @volume.setter
+    def volume(self, volume):
+        if not isinstance(volume, Cifti2Volume):
+            raise ValueError("You can only set a volume with a volume")
+        for i, v in enumerate(self):
+            if isinstance(v, Cifti2Volume):
+                break
+        else:
+            self.append(volume)
+            return
+        self[i] = volume
+
+    @volume.deleter
+    def volume(self):
+        for i, v in enumerate(self):
+            if isinstance(v, Cifti2Volume):
+                break
+        else:
+            raise ValueError("No Cifti2Volume element")
+        del self[i]
+
+    @property
+    def brain_models(self):
+        for p in self:
+            if isinstance(p, Cifti2BrainModel):
+                yield p
+
+    def _to_xml_element(self):
+        if self.applies_to_matrix_dimension is None:
+            raise CIFTI2HeaderError(
+                'MatrixIndicesMap element requires to be applied to at least 1 dimension'
+            )
+
+        mat_ind_map = xml.Element('MatrixIndicesMap')
+        dims_as_strings = [str(dim) for dim in self.applies_to_matrix_dimension]
+        mat_ind_map.attrib['AppliesToMatrixDimension'] = ','.join(dims_as_strings)
+        for key in ['IndicesMapToDataType', 'NumberOfSeriesPoints', 'SeriesExponent',
+                    'SeriesStart', 'SeriesStep', 'SeriesUnit']:
+            attr = _underscore(key)
+            value = getattr(self, attr)
+            if value is not None:
+                mat_ind_map.attrib[key] = str(value)
+        for map_ in self:
+            mat_ind_map.append(map_._to_xml_element())
+
+        return mat_ind_map
+
+
+class Cifti2Matrix(xml.XmlSerializable, collections.MutableSequence):
+    def __init__(self):
+        self._mims = []
+        self.metadata = None
+
+    @property
+    def metadata(self):
+        return self._meta
+
+    @metadata.setter
+    def metadata(self, meta):
+        """ Set the metadata for this Cifti2Header
+
+        Parameters
+        ----------
+        meta : Cifti2MetaData
+
+        Returns
+        -------
+        None
+        """
+        self._meta = _value_if_klass(meta, Cifti2MetaData)
+
+    def __setitem__(self, key, value):
+        if not isinstance(value, Cifti2MatrixIndicesMap):
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims[key] = value
+
+    def __getitem__(self, key):
+        return self._mims[key]
+
+    def __delitem__(self, key):
+        del self._mims[key]
+
+    def __len__(self):
+        return len(self._mims)
+
+    def insert(self, index, value):
+        if not isinstance(value, Cifti2MatrixIndicesMap):
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims.insert(index, value)
+
+    def _to_xml_element(self):
+        if (len(self) == 0 and self.metadata is None):
```

Comment:
Oh, as for a second API later - in workbench, we handled this by having separate functions for getting or setting the XML for the old and new API objects.  The ones for the older object translates it to/from the new object (which I did the lazy way by serializing to an XML byte array...) and otherwise relies on the equivalent function for the new API object.

Another suggested helper member function of the XML object:

`getNumberOfDimensions()` - searches the AppliesToMatrixDimension elements for the max value, return 1 more than the max

Those 3 functions should make it easy to deal with the XML object as if it were a simple array of mapping types *indexed by dimension*, which is really what its semantics are meant to be (aside from metadata and version).  Other helpers would mainly belong on the mapping objects, rather than on the XML object, and they are not as impactful.

Of course, I am using c++ idioms, so if it makes more sense from a python perspective to provide these operations a different way, by all means.